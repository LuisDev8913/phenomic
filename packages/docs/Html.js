import React from "react"
import Head from "react-helmet"

export type HtmlPropsType = {
  body: React$Element<*>,
  state?: React$Element<*>,
  script: React$Element<*>,
}

const Html = (props: HtmlPropsType) => {
  const helmet = Head.renderStatic()
  return (
    <html {...helmet.htmlAttributes.toComponent()}>
      <head>
        {helmet.base.toComponent()}
        {helmet.title.toComponent()}
        {helmet.meta.toComponent()}
        {helmet.link.toComponent()}
        {helmet.style.toComponent()}
        {helmet.script.toComponent()}
        {helmet.noscript.toComponent()}
      </head>
      <body {...helmet.bodyAttributes.toComponent()}>
        {props.body}
        {props.state}
        {props.script}
      </body>
    </html>
  )
}

export default Html
