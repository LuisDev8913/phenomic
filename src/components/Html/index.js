// @flow

import React from "react"
import { renderToString } from "react-dom/server"
import Helmet from "react-helmet"

type Props = {
  css: Array<string>,
  js: Array<string>,
  renderBody: () => React$Element<any>,
  renderScript: () => React$Element<any>,
}

const defaultHtmlAttributes = {
  "lang": "en",
}
const defaultMeta = [
  // <meta charset="utf-8" />
  { "charset": "utf-8" },
  // <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  { "http-equiv": "X-UA-Compatible", "content": "IE=edge" },
  // <meta name="viewport" content="width=device-width, initial-scale=1" />
  { "name": "viewport", "content": "width=device-width, initial-scale=1" },
]

const Html = (props: Props) => {
  // Inject default html metas before
  // Those need to be rendered somehow otherwise Helmet won't consider those
  renderToString(
    <Helmet
      htmlAttributes={ defaultHtmlAttributes }
      meta={ defaultMeta }
      link={ [
        ...props.css.map((file) => ({ rel: "stylesheet", href: file })),
      ] }
      script={ [
        ...props.js.map((file) => ({ src: file })),
      ] }
    />
  )

  // render body
  const body = props.renderBody()
  // rewind html metas
  const head = Helmet.rewind()

  // <!doctype html> is automatically prepended
  return (
    <html { ...head.htmlAttributes.toComponent() }>
      <head>
        { head.base.toComponent() }
        { head.title.toComponent() }
        { head.meta.toComponent() }
        { head.link.toComponent() }
      </head>
      <body>
        { body }
        { props.renderScript() }
        { head.script.toComponent() }
      </body>
    </html>
  )
}

export default Html
