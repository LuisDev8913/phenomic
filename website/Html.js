import * as React from "react";
import Head from "react-helmet";
import { StyleSheet } from "react-native-web";

const Html = ({ App, render }: PhenomicHtmlPropsType) => {
  const { Main, State, Script, Style } = render(<App />);
  const helmet = Head.renderStatic();
  return (
    <html {...helmet.htmlAttributes.toComponent()}>
      <head>
        {helmet.meta.toComponent()}
        {helmet.title.toComponent()}
        {helmet.base.toComponent()}
        <Style />
        {StyleSheet.getStyleSheets().map(({ id, textContent }, i) => (
          <style
            key={i}
            id={id}
            dangerouslySetInnerHTML={{ __html: textContent }}
          />
        ))}
        {helmet.link.toComponent()}
        {helmet.style.toComponent()}
        {helmet.script.toComponent()}
        {helmet.noscript.toComponent()}
      </head>
      <body {...helmet.bodyAttributes.toComponent()}>
        <Main />
        <State />
        <Script />
      </body>
    </html>
  );
};

export default Html;
