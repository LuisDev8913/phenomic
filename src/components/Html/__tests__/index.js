import React from "react"
import { createRenderer } from "react-addons-test-utils"

import Html from "../index.js"

test("should render Html componnent", () => {
  const renderer = createRenderer()
  renderer.render(
    /* eslint-disable react/jsx-no-bind */
    <Html
      css={ [ "test.css" ] }
      js={ [ "test.css" ] }
      renderBody={ () => <div /> }
      renderScript={ () => <script /> }
    />
  )
  expect(renderer.getRenderOutput()).toMatchSnapshot()
})
