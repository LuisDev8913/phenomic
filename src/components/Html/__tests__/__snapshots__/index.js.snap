exports[`test should render Html component 1`] = `
<html
  lang="en">
  <head>
    <title
      data-react-helmet={true} />
    <meta
      charSet="utf-8" />
    <meta
      content="IE=edge"
      httpEquiv="X-UA-Compatible" />
    <meta
      content="width=device-width, initial-scale=1"
      name="viewport" />
    <link
      href="test.css"
      rel="stylesheet" />
  </head>
  <body>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div />",
        }
      }
      id="phenomic" />
    <script
      phenomicStuff={true} />
    <script
      src="test.js" />
  </body>
</html>
`;
